File: src-tauri/src/commands/agent.rs
use crate::models::result::CommandResult;
use crate::services::agent_service::{
    self, Agent, AgentListData, CreateAgentPayload, DeleteAgentData, UpdateAgentPayload,
};

#[tauri::command]
pub async fn start_agent(id: String) -> CommandResult<Agent> {
    match agent_service::start_agent(id).await {
        Ok(agent) => CommandResult::ok(agent),
        Err(error) => CommandResult::err(error),
    }
}

#[tauri::command]
pub async fn stop_agent(id: String) -> CommandResult<Agent> {
    match agent_service::stop_agent(id).await {
        Ok(agent) => CommandResult::ok(agent),
        Err(error) => CommandResult::err(error),
    }
}

#[tauri::command]
pub async fn list_agents(search: Option<String>) -> CommandResult<AgentListData> {
    match agent_service::list_agents(search).await {
        Ok(data) => CommandResult::ok(data),
        Err(error) => CommandResult::err(error),
    }
}

#[tauri::command]
pub async fn create_agent(payload: CreateAgentPayload) -> CommandResult<Agent> {
    match agent_service::create_agent(payload).await {
        Ok(agent) => CommandResult::ok(agent),
        Err(error) => CommandResult::err(error),
    }
}

#[tauri::command]
pub async fn update_agent(payload: UpdateAgentPayload) -> CommandResult<Agent> {
    match agent_service::update_agent(payload).await {
        Ok(agent) => CommandResult::ok(agent),
        Err(error) => CommandResult::err(error),
    }
}

#[tauri::command]
pub async fn delete_agent(id: String) -> CommandResult<DeleteAgentData> {
    match agent_service::delete_agent(id).await {
        Ok(data) => CommandResult::ok(data),
        Err(error) => CommandResult::err(error),
    }
}

File: src/main/ipc/agent.ts
import { ipcMain } from "electron";
import { IPC } from "../../shared/ipc-channels.js";
import { ok, err } from "../models/result.js";
import { toAppError } from "./utils.js";
import {
  listAgents,
  createAgent,
  updateAgent,
  deleteAgent,
  startAgent,
  stopAgent,
} from "../services/agent-service.js";
import type {
  CreateAgentPayload,
  UpdateAgentPayload,
} from "../services/agent-service.js";

export function registerAgentHandlers(): void {
  ipcMain.handle(IPC.LIST_AGENTS, async (_event, payload?: { search?: string }) => {
    try { return ok(await listAgents(payload?.search)); }
    catch (error) { return err(toAppError(error)); }
  });

  ipcMain.handle(IPC.CREATE_AGENT, async (_event, payload: CreateAgentPayload) => {
    try { return ok(await createAgent(payload)); }
    catch (error) { return err(toAppError(error)); }
  });

  ipcMain.handle(IPC.UPDATE_AGENT, async (_event, payload: UpdateAgentPayload) => {
    try { return ok(await updateAgent(payload)); }
    catch (error) { return err(toAppError(error)); }
  });

  ipcMain.handle(IPC.DELETE_AGENT, async (_event, payload: { id: string }) => {
    try { return ok(await deleteAgent(payload.id)); }
    catch (error) { return err(toAppError(error)); }
  });

  ipcMain.handle(IPC.START_AGENT, async (_event, payload: { id: string }) => {
    try { return ok(await startAgent(payload.id)); }
    catch (error) { return err(toAppError(error)); }
  });

  ipcMain.handle(IPC.STOP_AGENT, async (_event, payload: { id: string }) => {
    try { return ok(await stopAgent(payload.id)); }
    catch (error) { return err(toAppError(error)); }
  });
}
